doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,215 @@
+## Problem Description
+
+We require a public API so that a developer can register callbacks for certain
+events like the arrival of a sample.
+
+## Name of the design element
+
+ - ReactAl = React And Listen
+ - Reactor
+ - Eventler
+ - LiAR - Listen And React
+
+
+## Terminology
+
+ - **event driven** a one time reaction which is caused directly by an event.
+      Example: a new sample has been delivered to a subscriber.
+ - **state driven** a repeating reaction which is caused by a specific structural 
+      state until the state changes.
+      Example: a subscriber has stored samples which were not inspected by the user.
+ - **robust API** we use this term to describe an API which is nearly impossible 
+      to misuse. Important in concurrent code since a misuse can lead to race 
+      conditions, Heisenbugs and in general hard to debug problems.
+ - **robust API feature** a feature we only support to increase the robustness 
+      of the API and bugs caused by misuse.
+ - (Reactor pattern)[https://en.wikipedia.org/wiki/Reactor_pattern]
+ - (Condition Variable)[https://en.wikipedia.org/wiki/Monitor_(synchronization)#Condition_variables_2]
+ 
+## Design
+
+### General
+The usage should be similar to the _WaitSet_ with a key difference - it should 
+be **event driven** and not a mixture of event and state driven, depending on
+which event is attached, like in the _WaitSet_.
+
+It should have the following behavior:
+
+ - Whenever an event occurs the corresponding callback should be called **once**
+    immediately.
+ - If an event occurs multiple times before the callback was called, the callback
+    should be called **once**.
+ - If an event occurs while the callback is being executed the callback should be 
+    called again **once**.
+
+The following use cases and behaviors should be implemented.
+
+ - The API and ReactAl should be robust this means that it should be impossible 
+    to use the API in the wrong way. Since ReactAl is working concurrently a 
+    wrong usage could lead to race conditions, extremely hard to debug bug reports 
+    (HeisenBugs) and can be frustrating to the user.
+    We list here features marked with [robust] which are only supported to 
+    increase this kind of the robustness.
+
+ - concurrently: attaching a callback at any time from anywhere. This means in particular
+    - Attaching a callback from within a callback.
+    - [robust] Updating a callback from within the same callback 
+      ```cpp
+      void sampleReceived2(iox::popo::UntypedSubscriber & subscriber) {}
+
+      void sampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+        myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, sampleReceived2);
+      }
```

Comment:
I documented the following: One callback per event of an object. Settings a callback when a callback is already attached overrides the callback with the new one.

The rational here is:
 - Overriding callbacks is no problem for us and allowing it per default reduces the error handling on our and the user side
 - It is exactly what my intuition expects when I call an `attachEvent` with a new callback.
 - We could deny it but this could also be a source of bugs due to a type or a c&p error. An error message could clear things somehow up, if the bug is tested. But as a compromise, how about a tracing message every time the callback is updated? Then the user can take a look at the logs and knows what's going on!

I would not use the `AttachmentPolicy` idea since for me it is intuitively clear that `attachEvent` assigns the callback. 

But maybe we should rename the method into `assignCallback` then it is more clear that a callback is assigned and it is like assigning a number.

```cpp
int a = 5; 
// and now assign 8
a = 8; // and everyone knows that the old value 5 is gone
```

Reply:
I agree. Assign makes it clearer for me compared to attach.

Reply:
+1 for assign